### ISSUE TITLE
gf, :find in terminal buffer looks for wrong path

### ISSUE THREAD

**Issue by still-dreaming-1** [neovim/neovim#4299]
- Neovim version: NVIM 0.1.3-dev
- Operating system: Linux Mint 17.3 64 bit Cinnamon
- Terminal emulator:
  Not sure what you mean, it is the Neovim terminal emulator. But I am using the default terminal program of Linux Mint, which is gnome-terminal. Once that is running, my shell is set to zsh. I have prezto installed for zsh. Once I type :te in Neovim, it does load up zsh whith my settings including the prezto stuff.
### Actual behaviour

After typing `gf` from Neovim's terminal emulator, it searches using the wrong path, so an error like the following shows up.

> zsh:1: no such file or directory: /bin/www/ACSmanager/l/experiment.php
### Expected behaviour

After pressing `gf`, it should navigate to that file using the same path assumptions that would happen in a normal buffer, or perhaps it should use the current working directory of the terminal. I have the current working directory in Neovim the same as the current directory shown in the terminal buffer, but it is still not using that directory to find the file.
### Steps to reproduce

Add a new untracked file to a git repository. Enter git status into the terminal emulator inside Neovim. It shows the file as untracked. An example of a path that can show up:
`www/ACSmanager/l/experiment.php`
 Go to normal mode. Move cursor to file. Type `gf`
The following error shows up: 

> zsh:1: no such file or directory: /bin/www/ACSmanager/l/experiment.php

**Comment by mhinz** [comment]
I can see why this is happening. Your terminal buffer probably runs `/bin/zsh` and `:set path` contains ".". So on `gf`, Neovim tries the wrong "current directory" `/bin` instead of the actual working directory.

**Comment by justinmk** [comment]
[OSC 7](https://github.com/neovim/neovim/issues/2252) could help with this, though we still need a solution for the terminals that don't send that.

**Comment by still-dreaming-1** [comment]
Just an FYI. My package manager showed a new version of Neovim ready to install. I installed it, and this problem still exists. Not that I was expecting it be fixed already. Don't know if this is helpful, but here you go:
`nvim --version                                                                                                                                       ⮂  
NVIM 0.1.3-dev
Build type: RelWithDebInfo
Compilation: /usr/bin/x86_64-linux-gnu-gcc -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -DDISABLE_LOG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector --param ssp-buffer-size=4 -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -D_GNU_SOURCE -I/build/neovim-c5eGGt/neovim-0.1.2ubuntu1+git201602201552+2289+17~ubuntu14.04.1/build/config -I/build/neovim-c5eGGt/neovim-0.1.2ubuntu1+git201602201552+2289+17~ubuntu14.04.1/src -I/build/neovim-c5eGGt/neovim-0.1.2ubuntu1+git201602201552+2289+17~ubuntu14.04.1/.deps/usr/include -I/build/neovim-c5eGGt/neovim-0.1.2ubuntu1+git201602201552+2289+17~ubuntu14.04.1/.deps/usr/include -I/usr/include/luajit-2.0 -I/build/neovim-c5eGGt/neovim-0.1.2ubuntu1+git201602201552+2289+17~ubuntu14.04.1/.deps/usr/include -I/build/neovim-c5eGGt/neovim-0.1.2ubuntu1+git201602201552+2289+17~ubuntu14.04.1/.deps/usr/include -I/build/neovim-c5eGGt/neovim-0.1.2ubuntu1+git201602201552+2289+17~ubuntu14.04.1/.deps/usr/include -I/build/neovim-c5eGGt/neovim-0.1.2ubuntu1+git201602201552+2289+17~ubuntu14.04.1/.deps/usr/include -I/usr/include -I/build/neovim-c5eGGt/neovim-0.1.2ubuntu1+git201602201552+2289+17~ubuntu14.04.1/build/src/nvim/auto -I/build/neovim-c5eGGt/neovim-0.1.2ubuntu1+git201602201552+2289+17~ubuntu14.04.1/build/include
Compiled by buildd@lgw01-04`

**Comment by wsdjeg** [comment]
I can reproduce this issue : 
1. open neovim 
2. type `:te`
3. type into `ls` will list all the files
4. type `<c-\><c-n>` 
5. move cursor to one file,and type `gf`

**Comment by justinmk** [comment]
This is low priority. Unlikely to be fixed soon unless someone sends a PR.

**Comment by still-dreaming-1** [comment]
@justinmk Doesn't this bug strike to the very core of why people would want to use Neovim over Vim? There must be something I am missing. Why am I the first person being bothered by this bug? Is there a workaround I can try in the mean time? When I say workaround, I don't mean another way to open files, I mean a way to make gf work in a terminal buffer. Is everyone else using bash or something?

**Comment by fmoralesc** [comment]
> Why am I the first person being bothered by this bug?

It's probably a combination of things. I use fish and the error is even worse:
~~~
fish: Unknown command 'filename'
fish: filename
      ^

[Process exited 127
~~~
Meanwhile, perhaps you can map gf to something like
~~~ vim
nmap <buffer> gf :e <cfile><cr>
~~~
?

**Comment by justinmk** [comment]
> Doesn't this bug strike to the very core of why people would want to use Neovim over Vim? 

There's a wide between "no terminal" and "perfectly integrated terminal". Neovim doesn't claim the latter, but offers something in the middle. Emacs claims the latter and fails in ways that leave users in various states of satisfaction.

Neovim's terminal integration is better than none at all. The fact that some built-in concepts don't yet translate to that integration does not strike me as a risk to "the very core" of Neovim's purpose. E.g., I can copy, paste, and search using tmux panes instead of Neovim buffers, but that's objectively _less functional_. The fact that `gf` doesn't yet work is no worse than the situation with tmux or whatever.

I do think we should consider delegating `gf` to a provider or a hook at some point.

**Comment by still-dreaming-1** [comment]
@justinmk My point is that it takes away the appeal of switching to Neovim as there are already things like ConqueTerm and VimShell that work better than this. I am not complaining about the state of Neovim, I know it is a work in progress. I am just saying something like this should be considered one of the more high priority bugs to fix for the reasons just mentioned. I am still at a loss as to why other people were not already reporting this problem. Zsh is a very popular shell and the whole point of integrating a Terminal emulator into Neovim is not to get access to a terminal or shell, but to have it run in  Vim buffer and therefore have the advantages of working from Vim. Anyway, I am not going to keep arguing this point further, as I feel I have already expressed myself well.

@fmoralesc That was a good idea for a workaround. I was able to remove some, but not all of the limitations of that workaround by using the normal gf when it is not running from a buffer:

`nnoremap gf :call Gf()<enter>                                                                                                                            
function! Gf()                                                                                                                                           
    if &buftype ==# "terminal"                                                                                                                           
        :e <cfile>                                                                                                                                       
    else                                                                                                                                                 
        normal! gf                                                                                                                                       
    endif                                                                                                                                                
endfunction`

**Comment by fmoralesc** [comment]
Maybe this might help:
~~~ vim
function! s:term_gf()
    let procid = matchstr(bufname(""), '\(://.*/\)\@<=\(\d\+\)')
    let proc_cwd = resolve('/proc/'.procid.'/cwd')
    exe 'lcd '.proc_cwd
    exe 'e <cfile>'
endfunction

au TermOpen * nmap <buffer> gf :call <SID>term_gf()<cr>
~~~
This assumes the typical naming scheme for terminal buffers, which includes the child process number.

**Comment by still-dreaming-1** [comment]
@fmoralesc I wish this one liner had worked though. Not sure why it is not working:

`"autocmd BufReadPost terminal nnoremap <buffer> gf :e <cfile><enter>`

**Comment by fmoralesc** [comment]
The issue there is `BufReadPost` doesn't take a buffer type as argument (I assume that's what you intended by using "terminal" after it). Better use `TermOpen *`, as in my last variant.

**Comment by justinmk** [comment]
Even with TermOpen, `<cfile>` won't know the terminal's current directory. 

But the `/proc/` approach looks like a nice workaround! Maybe we could use that workaround in `find_file_name_in_path()`. Anyone want to look at that?

For OSC7 we are waiting on libvterm to add hooks for unhandled control sequences: https://blueprints.launchpad.net/libvterm/+spec/unrecognised-fallback

> there are already things like ConqueTerm and VimShell that work better than this

Hmm. I tried them both, and must say I prefer `:te` by miles.

**Comment by still-dreaming-1** [comment]
@fmoralesc That workaround works really well! It's totes digi! :D Thanks!

**Comment by fmoralesc** [comment]
The `/proc/` method is basically what people used before OSC 7 -- I was simply too dumb to figure out how OSC 7 is supposed to work, so I had to look elsewhere ;) 

The whole approach has a big problem, though (other than cross-platform support). In a typical `:terminal` session, you might have
~~~ sh
$ cd dirA
$ ls
subdirA subdirB
$ cd subdirA
$ ls
afile anotherfile
$ cd ..
~~~
What happens when the user moves over "afile" and presses gf? The command will fail. So no matter what we do, we can only solve the behavior of `gf` for paths that are relative to the current working directory (to be fair, that's probably the most common scenario, so it's not a bad thing). For solving the general case we would need a procedure that could look in the word context and inferred the relevant path, or that looked upwards and downwards in the directory tree to look for the file. It gets complicated quickly.

**Comment by still-dreaming-1** [comment]
@fmoralesc As long as gf at least assumes the current working directory of the terminal that would be a huge improvement. Getting it to work from previous output when the user was in a different directory would be pretty nice to have. But gf doesn't normally work that way anyway when you are inside a file buffer. It doesn't magically know where you want to open that path from, it has to make assumptions about that. In the case of a terminal buffer at least the information to know what directory was the context for different output at least exists somewhere so it would be nice for that to work. That would be a great user experience.

**Comment by justinmk** [comment]
> The whole approach has a big problem, though (other than cross-platform support)

Isn't `/proc/` on BSD and linux? Though not OSX, but apparently [libproc.h](http://stackoverflow.com/a/33852283/152142) is available.

> we would need a procedure that could look in the word context and inferred the relevant path

Which is fragile. I also seriously doubt that VimShell, Conque or even Emacs can handle that.

**Comment by fmoralesc** [comment]
> As long as gf at least assumes the current working directory of the terminal that would be a huge improvement. 

@still-dreaming-1 My s:term_gf() function does that, at least.

**Comment by BohrShaw** [comment]
I can reproduce this issue with:

`nvim -u NORC +'set path+=~/.config | term' +'find nvim/init.vim'`

on both zsh and bash as the default shell. #4572

**Comment by ticki** [comment]
With the latest nightly (and zsh as shell) `gf` opens a new terminal buffer which errors with code 127, making `gf` unusable with error output. This is a major annoyance for me. The work-arounds suggested in this thread doesn't seem to work for me...

**Comment by mattsmithdatera** [comment]
This issue is still occurring with zsh as of today's latest dev build NVIM v0.1.5-957-g7df9dd3

**Comment by jamessan** [comment]
> This issue is still occurring with zsh as of today's latest dev build NVIM v0.1.5-957-g7df9dd3

Yes, because the issue is still open.  There's no need to continue confirming a confirmed bug.

If people want to help, digging into the code is the approach to take.

**Comment by mattsmithdatera** [comment]
Sorry man.  Just wanted to provide helpful information for folks that might stumble across it.  If it helps, the workaround I've been using is the one shown by @still-dreaming-1 a few posts up in this thread.

**Comment by BohrShaw** [comment]
I just removed `.,,` in the `path` option to make `:find` on a terminal buffer work.

**Comment by still-dreaming-1** [comment]
I've only tried this in a couple places, but it works well so far. Thanks!
